File: lake/amoro/ams/server/src/main/java/com/netease/arctic/server/table/TableService.java
/*
 * Licensed to the Apache Software Foundation (ASF) under one
 * or more contributor license agreements.  See the NOTICE file
 * distributed with this work for additional information
 * regarding copyright ownership.  The ASF licenses this file
 * to you under the Apache License, Version 2.0 (the
 * "License"); you may not use this file except in compliance
 * with the License.  You may obtain a copy of the License at
 *
 *     http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 */

package com.netease.arctic.server.table;

import com.netease.arctic.ams.api.BlockableOperation;
import com.netease.arctic.ams.api.Blocker;
import com.netease.arctic.ams.api.TableIdentifier;
import com.netease.arctic.server.catalog.CatalogService;

import java.util.List;
import java.util.Map;

public interface TableService extends CatalogService, TableManager {

  void initialize();

  /**
   * create table metadata
   *
   * @param tableMeta table metadata info
   */
  void createTable(String catalogName, TableMetadata tableMeta);

  /**
   * load the table metadata
   *
   * @param tableIdentifier table id
   * @return table metadata info
   */
  TableMetadata loadTableMetadata(TableIdentifier tableIdentifier);

  /**
   * delete the table metadata
   *
   * @param tableIdentifier table id
   * @param deleteData      if delete the external table
   */
  void dropTableMetadata(TableIdentifier tableIdentifier, boolean deleteData);

  /**
   * load arctic databases name
   *
   * @return databases name list
   */
  List<String> listDatabases(String catalogName);

  /**
   * Load all managed tables.
   * Managed tables means the tables which are managed by AMS, AMS will watch their change and make them health.
   *
   * @return {@link ServerTableIdentifier} list
   */
  List<ServerTableIdentifier> listManagedTables();

  /**
   * Load all managed tables.
   * Managed tables means the tables which are managed by AMS, AMS will watch their change and make them health.
   *
   * @return {@link ServerTableIdentifier} list
   */
  List<ServerTableIdentifier> listManagedTables(String catalogName);

  /**
   * Load table identifiers by server catalog
   *
   * @return {@link TableIdentifier} list
   */
  List<TableIdentifier> listTables(String catalogName, String dbName);

  /**
   * create arctic database
   */
  void createDatabase(String catalogName, String dbName);

  /**
   * drop arctic database
   */
  void dropDatabase(String catalogName, String dbName);

  /**
   * load all arctic table metadata
   *
   * @return table metadata list
   */
  List<TableMetadata> listTableMetas();

  /**
   * load arctic table metadata
   *
   * @return table metadata list
   */
  List<TableMetadata> listTableMetas(String catalogName, String database);

  /**
   * check the table is existed
   *
   * @return True when the table is existed.
   */
  boolean tableExist(TableIdentifier tableIdentifier);

  /**
   * blocker operations
   *
   * @return the created blocker
   */
  Blocker block(TableIdentifier tableIdentifier, List<BlockableOperation> operations, Map<String, String> properties);

  /**
   * release the blocker
   */
  void releaseBlocker(TableIdentifier tableIdentifier, String blockerId);

  /**
   * renew the blocker
   *
   * @return expiration time
   */
  long renewBlocker(TableIdentifier tableIdentifier, String blockerId);

  /**
   * getRuntime blockers of table
   *
   * @return block list
   */
  List<Blocker> getBlockers(TableIdentifier tableIdentifier);
}

File: lake/amoro/ams/dashboard/src/services/table.service.ts
/*
  * Licensed to the Apache Software Foundation (ASF) under one
  * or more contributor license agreements.  See the NOTICE file
  * distributed with this work for additional information
  * regarding copyright ownership.  The ASF licenses this file
  * to you under the Apache License, Version 2.0 (the
  * "License"); you may not use this file except in compliance
  * with the License.  You may obtain a copy of the License at
  *
  *     http://www.apache.org/licenses/LICENSE-2.0
  *
  * Unless required by applicable law or agreed to in writing, software
  * distributed under the License is distributed on an "AS IS" BASIS,
  * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
  * See the License for the specific language governing permissions and
  * limitations under the License.
  */

// import { IOptions } from '@/types/common.type'
import { ICatalogItem, IMap } from '@/types/common.type'
import request from '@/utils/request'

export function getCatalogList(): Promise<ICatalogItem[]> {
  return request.get('ams/v1/catalogs')
}
export function getDatabaseList(params: {
  catalog: string
  keywords: string
}): Promise<string[]> {
  const { catalog, keywords } = params
  return request.get(`ams/v1/catalogs/${catalog}/databases`, { params: { keywords } })
}

export function getTableList(params: {
  catalog: string
  db: string
  keywords: string
}) {
  const { catalog, db, keywords } = params
  return request.get(`ams/v1/catalogs/${catalog}/databases/${db}/tables`, { params: { keywords } })
}

// get tables detail
export function getTableDetail(
  { catalog = '' as string, db = '' as string, table = '' as string, token = '' as string }
) {
  return request.get(`ams/v1/tables/catalogs/${catalog}/dbs/${db}/tables/${table}/details`, { params: { token } })
}

export function getHiveTableDetail(
  { catalog = '' as string, db = '' as string, table = '' as string }
) {
  return request.get(`ams/v1/tables/catalogs/${catalog}/dbs/${db}/tables/${table}/hive/details`)
}

export function getUpgradeStatus(
  { catalog = '' as string, db = '' as string, table = '' as string }
) {
  return request.get(`ams/v1/tables/catalogs/${catalog}/dbs/${db}/tables/${table}/upgrade/status`)
}
// get partions table
export function getPartitionTable(
  params: {
    catalog: string
    db: string,
    table: string,
    page: number
    pageSize: number
    token: string
  }
) {
  const { catalog, db, table, page, pageSize, token } = params
  return request.get(`ams/v1/tables/catalogs/${catalog}/dbs/${db}/tables/${table}/partitions`, { params: { page, pageSize, token } })
}

// get partions
export function getPartitions(
  params: {
    catalog: string
    db: string,
    table: string,
    page: number
    pageSize: number
    token: string
  }
) {
  const { catalog, db, table, page, pageSize, token } = params
  return request.get(`ams/v1/tables/catalogs/${catalog}/dbs/${db}/tables/${table}/partitions`, { params: { page, pageSize, token } })
}
// get partions-files
export function getPartitionFiles(
  params: {
    catalog: string
    db: string,
    table: string,
    partition: string | null,
    page: number
    pageSize: number
    token: string
  }
) {
  const { catalog, db, table, partition, page, pageSize, token } = params
  return request.get(`ams/v1/tables/catalogs/${catalog}/dbs/${db}/tables/${table}/partitions/${partition}/files`, { params: { page, pageSize, token } })
}
// get Transactions
export function getTransactions(
  params: {
    catalog: string
    db: string,
    table: string,
    page: number
    pageSize: number
    token: string
  }
) {
  const { catalog, db, table, page, pageSize, token } = params
  return request.get(`ams/v1/tables/catalogs/${catalog}/dbs/${db}/tables/${table}/transactions`, { params: { page, pageSize, token } })
}

// get TransactionId detail
export function getDetailByTransactionId(
  params: {
    catalog: string
    db: string,
    table: string,
    transactionId: string,
    page: number
    pageSize: number
    token: string
  }
) {
  const { catalog, db, table, transactionId, page, pageSize, token } = params
  return request.get(`ams/v1/tables/catalogs/${catalog}/dbs/${db}/tables/${table}/transactions/${transactionId}/detail`, { params: { page, pageSize, token } })
}
// get operations
export function getOperations(
  params: {
    catalog: string
    db: string,
    table: string,
    page: number
    pageSize: number
    token: string
  }
) {
  const { catalog, db, table, page, pageSize, token } = params
  return request.get(`ams/v1/tables/catalogs/${catalog}/dbs/${db}/tables/${table}/operations`, { params: { page, pageSize, token } })
}
// get optimizes
export function getOptimizes(
  params: {
    catalog: string
    db: string,
    table: string,
    page: number
    pageSize: number
    token: string
  }
) {
  const { catalog, db, table, page, pageSize, token } = params
  return request.get(`ams/v1/tables/catalogs/${catalog}/dbs/${db}/tables/${table}/optimizing-processes`, { params: { page, pageSize, token } })
}

export function upgradeHiveTable(
  { catalog = '' as string, db = '' as string, table = '' as string, properties = {} as IMap<string>, pkList = [] as IMap<string>[] }
) {
  return request.post(`ams/v1/tables/catalogs/${catalog}/dbs/${db}/tables/${table}/upgrade`, {
    properties,
    pkList
  })
}

export function getUpgradeProperties() {
  return request.get('ams/v1/upgrade/properties')
}
